pages/exercises/[lessonSlug].tsx

```diff
@@ -70,31 +71,34 @@ const Exercises: React.FC<QueryDataProps<GetExercisesQuery>> = ({
       explanation: exercise.explanation || '',
       userAnswer: userAnswers[exercise.id] ?? null
     }))
-
-  const exercise = currentExercises[exerciseIndex]
+    .filter(
+      exercise => !hideAnswered || exercise.userAnswer !== exercise.answer
+    )
 
   return (
     <Layout title={currentLesson.title}>
-      {exercise ? (
+      {solvingExercise ? (
         <Exercise
-          key={exerciseIndex}
-          exercise={exercise}
-          setExerciseIndex={setExerciseIndex}
           lessonTitle={currentLesson.title}
-          hasPrevious={exerciseIndex > 0}
-          hasNext={exerciseIndex < currentExercises.length - 1}
-          submitUserAnswer={(userAnswer: string) => {
-            setUserAnswers({ ...userAnswers, [exercise.id]: userAnswer })
+          exercises={currentExercises}
+          userAnswers={userAnswers}
+          onExit={localUserAnswers => {
+            setUserAnswers({ ...userAnswers, ...localUserAnswers })
+            setSolvingExercise(false)
+          }}
+          submitUserAnswer={(exerciseId, userAnswer) => {
             addExerciseSubmission({
-              variables: { exerciseId: exercise.id, userAnswer }
+              variables: { exerciseId, userAnswer }
```

Comment:
We no longer update the `userAnswers` locally after each submit. Instead, we update the server exercise submissions and we update the `localUserAnswers` because we don't want updating `userAnswers` to change the `currentExercises` which would cause a rerender in a buggy way when the exercise filtering checkbox is checked (e.g. if we only show incomplete exercises then we complete an exercise, then we would increment the exercise index and the exercises array would be shorter by 1 after that exercise got complete, so to avoid this we don't update `userAnswers` until after the user exits the `Exercise` component).